### DISCUSSION TITLE
캡슐화가 무엇이고, 이를 통해 얻을 수 있는 장점이 무엇인가요?

### DISCUSSION THREAD

**Discussion by gyungchan-jo** [2024-woowacourse-study/level-interview#105] [OOP]
.

**Comment by jimi567** [comment]
캡슐화는 객체 내부의 상태를 보호하고 외부에서의 불필요한 접근을 막도록, 멤버 변수와 메서드를 숨기는 것을 의미한다. 캡슐화에는 중요한 두 가지 개념이 있다.
1. 정보 은닉
 객체의 상태를 외부로부터 숨기고, 외부에서는 객체에 접근할 수있는 메서드를 통해서만 접근할 수 있도록 한다. 이를 통해 객체의 내부 구현을 숨기고 외부와의 결합도를 낮출 수 있다.
 2. 접근 제어
 접근 제어자를 설정해, 객체의 상태에 대해 무분별한 접근을 방지하고, 데이터의 무결성을 보존한다.
 
캡슐화의 장점은
객체 간의 결합도를 낮추어, 객체의 내부 구현이 변경되더라도, 클라이언트 코드(외부 코드)에 영향을 최소화할 수 있도록 도와준다.
이를 통해, 안정성과 재사용성을 향상 시켜, 유지보수 측면에서 이점을 얻을 수 있다.

**Comment by nayonsoso** [comment]
캡슐화란, 데이터와 그 데이터를 처리하는 행위를 묶고, 외부에 드러내지 않는 것입니다.
캡슐화를 사용하면, 외부에서 행위의 원리를 알지 못한채로 행위를 '요청'을 하게 되는데요.
이를 통해 객체간의 협력을 단순하게 만들 수 있습니다.

예를 들어 자동차 게임에서 자동차를 움직이고 싶을 때,
자동차 객체의 내부를 알 필요 없이 move() 함수만 호출한다면 코드가 훨씬 간결해지겠죠.
이렇게 행위를 요청하는 것을 "객체에 메세지를 보낸다"라고도 말합니다.

캡슐화의 장점으로는, 코드의 변경을 줄여준다는 것이 있습니다.
외부에서 내부를 알지 못하고 요청만 하기 때문에,
내부 로직이 변하는 경우, 외부 코드는 변경하지 않아도 됩니다.
또한 내부의 값이 예상치 못하게 변하는 것을 막을 수 있다는 장점도 있습니다.

캡슐화를 지키기 위해서 필드를 private으로 두거나, setter를 만들지 않는 등의 노력을 할 수 있습니다.

**Comment by le2sky** [comment]
객체 내부의 세부적인 사항을 감추는 것을 캡슐화라고 합니다. 
캡슐화의 목적은 변경하기 쉬운 객체를 만드는 것입니다. 
캡슐화를 통해 객체 내부로의 접근을 제한하면 객체와 객체 사이의 결합도를 낮출 수 있기 때문에 설계를 좀 더 쉽게 변경할 수 있습니다.

**Comment by GIVEN53** [comment]
세부 사항을 외부에 노출 x
내부를 모르게, 요청하도록
접근제어자 활용
결합도 낮춤
쉬운 변경
